### ISSUE TITLE
BUG: Unable to define non-default field in a Maker subclass 

### ISSUE THREAD

**Issue by davidwaroquiers** [materialsproject/jobflow#13]
Hi Alex, hi Anubhav,

I am playing around with jobflow and have started to write some workflows (flows) with jobflow (I have some comments that would deserve a discussion on its own,let's see about that another time :)

**Describe the bug**
Unable to define non-default field in a Maker subclass. It is not a bug of jobflow per se but it can be annoying (there is one simple fix but it's also a bit annoying, see below, maybe there are other options ?).

**To Reproduce**
@dataclass
class MyMaker(Maker):
    mandatory_argument: int
    optional_argument: int = 5

Because "name" in the parent (Maker) class has a default value ("Maker"), you get a TypeError saying that you define a non-default argument (here "mandatory_argument") after a default argument (the one inherited from the Maker class).

One possibility would be to make name mandatory. Any other ideas ?

**Comment by utf** [comment]
Hi David,

Thanks very much for this bug report. This seems like a limitation of data classes and there isn't an obvious way around it (after reading https://stackoverflow.com/questions/51575931/class-inheritance-in-python-3-7-dataclasses).

I played around with a few approaches:
- Making name required and then overriding as an optional field in subclasses (doesn't work, raises the same error).
- Making the `Maker` class a non-dataclass with subclasses implemented as dataclasses (inheritance doesn't work properly).
- Combinations of using `field(default="Maker", init=False)` and `post_init` to handle things (couldn't get it to work satisfactorily).

Essentially, we want to treat `name` as an "abstract field" that subclasses should implement. However, there isn't support for abstract fields natively in python. This package [dataclass-abc](https://pypi.org/project/dataclass-abc/) aims to do this but it hasn't been updated in > 1 year.

Instead, this is the best solution I could come up with. The base maker class defines an abstract property `name`.

```python
@dataclass
class Maker(MSONable):
        
    def make(self, *args, **kwargs):
        raise NotImplementedError
    
    @property
    def name(self):
        raise NotImplementedError
```

Subclasses should then include name in their fields:

```python
@dataclass
class SubMaker(Maker):
    
    required: int
    name: str = "SubMaker"    
```

This then works as you'd expect.
```python
s = SubMaker(required=1)
s.name  # ==  "SubMaker"
```

All of this will be explained in the docstrings. What do you think?

P.s., if you're using jobflow I'd love to hear your thoughts/any comments. Perhaps we could arrange a zoom?

**Comment by davidwaroquiers** [comment]
Sounds like a reasonable workaround. You could even keep the default "Maker" by doing:

```
@dataclass
class Maker(MSONable):
        
    def make(self, *args, **kwargs):
        raise NotImplementedError
    
    @property
    def name(self):
        return "Maker"
```

instead of raising. Not sure if this is better though, probably good that the user/developer always assigns a name to his/her Maker.


Sure we can have a zoom meeting, maybe the week from 5 to 9 July ? I indeed have a few points. Overall things are working very nicely, it's more questions and maybe a few small ideas or changes to propose.

**Comment by utf** [comment]
Great, I'll go ahead and implement this. I've been developing the atomate2 workflows and run into a few bugs in jobflow. Some of these have been fixed, others are being tested in another branch. One of the big things I did recently was removing the `Schema` class in favour of just using standard pydantic models.

That week works for me, I'll send you an email.

**Comment by utf** [comment]
This is fixed in the v0.1 branch by db666979d84296569e06c228495de85c21fc4ee9.
